### ISSUE TITLE
Changing training Parameters 

### ISSUE THREAD

**Issue by kolmorgan** [CSUBioGroup/PGMG#8]
@CSUBioGroup Hi, can you please share the modified version with me or tell me how to modify the code to change the pharmacophore to a number like 13? Also, can you tell me exactly how to add the ANION type to the pharmacophore training? I'm not an expert in Python and DL/ML, and I don't know how to change them without causing errors or problems. Thank you so much!

**Comment by kolmorgan** [comment]
Hi, I have modified the `MAX_NUM_PP_GRAPHS=13` in smiles2ppgraph.py and added the anion type by making small modifications to the code, specifically lines 32 and 128-130 in smiles2ppgraph.py         
`orgin_phco = [0, 0, 0, 0, 0, 0, 0, 0, 0]
mapping = {'Aromatic': 1, 'Hydrophobe': 2, 'PosIonizable': 3,
                   'Acceptor': 4, 'Donor': 5, 'LumpedHydrophobe': 6, 'Anion':7,}
        phar_index = mapping.setdefault(phar, 8)` and line 49 in train_chembl_baseline.py 
`pp_v_dim": 8 + 1,  # dimension of pharmacophore embedding vectors.`
 However, when I retrained the model, an error appeared
`Traceback (most recent call last):
  File "train_chembl_baseline.py", line 516, in <module>
    avg_loss = train_fn(train_loader, model, optimizer, epoch, scheduler, beta, scaler=scaler)
  File "train_chembl_baseline.py", line 157, in train_fn
    x[:, :xx.shape[1], :] = xx
RuntimeError: The expanded size of the tensor (7) must match the existing size (8) at non-singleton dimension 2.  Target sizes: [128, 7, 7].  Tensor sizes: [128, 7, 8]`
After consulting a friend of mine, the error may be related to line 39 in train_chembl_baseline.py, which is `PP_TYPE_WEIGHT = [1.4891304347826086, 1.0, 8.058823529411764, 1.0378787878787878, 1.8026315789473686, 2.174603174603175,
                  17.125]`, I'm not sure if this array reflects the weights of the existing pharmacophore types. In order to continue, I assume that it is, and the order is in accordance with the order in mapping, so I supplemented the value of anion with the same as cation with the number 8.058823529411764, and for now it starts training successfully. However, I know that the value was not a proper one, so is there any way that I can get the right value for the addition of Anion type? Looking forward to your reply!

**Comment by e-yi** [comment]
Hi, @kolmorgan. Glad to know you have got things worked out! 😄 

Your assumption is correct—the array indeed reflects the weights of pharmacophore types. Essentially, what we've done is $$weight_i= \frac{max_j({count_j})}{{count_i}} $$ for each pharmacophore element type $i$. Here, $count$ represents the number of occurrences in the training set. To determine precise weights, you might consider sampling a number of molecules, let's say 100,000, and then count the occurrences for each type. Please refer to L128-L132 in `smiles2ppgraph.py` to get all pharmacophore features within a molecule. But I'd say using the weight of the cation is a good enough approximation.

**Comment by kolmorgan** [comment]
After trying many times, the training seems worked to some extent, error messages appeared during the training, including
` 56%|█████▋    | 5550/9841 [19:02<12:16,  5.83it/s]ERROR: Unexpected segmentation fault encountered in worker.
\00
 56%|█████▋    | 5550/9841 [19:02<14:43,  4.86it/s]
Traceback (most recent call last):
  File "train_chembl_baseline.py", line 516, in <module>
    avg_loss = train_fn(train_loader, model, optimizer, epoch, scheduler, beta, scaler=scaler)
  File "train_chembl_baseline.py", line 183, in train_fn
    accumulated_loss.backward()
  File "/home/user/anaconda3/envs/pgmg/lib/python3.8/site-packages/torch/_tensor.py", line 487, in backward
    torch.autograd.backward(
  File "/home/user/anaconda3/envs/pgmg/lib/python3.8/site-packages/torch/autograd/__init__.py", line 197, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File "/home/user/anaconda3/envs/pgmg/lib/python3.8/site-packages/torch/utils/data/_utils/signal_handling.py", line 66, in handler
    _error_if_any_worker_fails()
RuntimeError: DataLoader worker (pid 9595) is killed by signal: Segmentation fault. ` and 
`Traceback (most recent call last):
  File "train_chembl_baseline.py", line 516, in <module>
    avg_loss = train_fn(train_loader, model, optimizer, epoch, scheduler, beta, scaler=scaler)
  File "train_chembl_baseline.py", line 144, in train_fn
    inputs, input_mask, pp_graphs, mappings, targets, *others = [i.to('cuda:0') for i in batch_data]
  File "train_chembl_baseline.py", line 144, in <listcomp>
    inputs, input_mask, pp_graphs, mappings, targets, *others = [i.to('cuda:0') for i in batch_data]
  File "/home/user/anaconda3/envs/pgmg/lib/python3.8/site-packages/dgl/heterograph.py", line 5448, in to
    ret._graph = self._graph.copy_to(utils.to_dgl_context(device))
  File "/home/user/anaconda3/envs/pgmg/lib/python3.8/site-packages/dgl/heterograph_index.py", line 237, in copy_to
    return _CAPI_DGLHeteroCopyTo(self, ctx.device_type, ctx.device_id)
  File "dgl/_ffi/_cython/./function.pxi", line 293, in dgl._ffi._cy3.core.FunctionBase.__call__
  File "dgl/_ffi/_cython/./function.pxi", line 225, in dgl._ffi._cy3.core.FuncCall
  File "dgl/_ffi/_cython/./function.pxi", line 215, in dgl._ffi._cy3.core.FuncCall3
dgl._ffi.base.DGLError: [12:51:42] /opt/dgl/src/runtime/cuda/cuda_device_api.cc:303: Check failed: e == cudaSuccess || e == cudaErrorCudartUnloading: CUDA: invalid argument
Stack trace:
  [bt] (0) /home/user/anaconda3/envs/pgmg/lib/python3.8/site-packages/dgl/libdgl.so(dmlc::LogMessageFatal::~LogMessageFatal()+0x4f) [0x152a86595e3f]
  [bt] (1) /home/user/anaconda3/envs/pgmg/lib/python3.8/site-packages/dgl/libdgl.so(dgl::runtime::CUDADeviceAPI::CopyDataFromTo(void const*, unsigned long, void*, unsigned long, unsigned long, DLContext, DLContext, DLDataType)+0x8d) [0x152a86a6e20d]
  [bt] (2) /home/user/anaconda3/envs/pgmg/lib/python3.8/site-packages/dgl/libdgl.so(dgl::runtime::NDArray::CopyFromTo(DLTensor*, DLTensor*)+0x267) [0x152a868deae7]
  [bt] (3) /home/user/anaconda3/envs/pgmg/lib/python3.8/site-packages/dgl/libdgl.so(dgl::runtime::NDArray::CopyTo(DLContext const&) const+0xe8) [0x152a8691b398]
  [bt] (4) /home/user/anaconda3/envs/pgmg/lib/python3.8/site-packages/dgl/libdgl.so(dgl::aten::COOMatrix::CopyTo(DLContext const&) const+0x7d) [0x152a86a3d88d]
  [bt] (5) /home/user/anaconda3/envs/pgmg/lib/python3.8/site-packages/dgl/libdgl.so(dgl::UnitGraph::CopyTo(std::shared_ptr<dgl::BaseHeteroGraph>, DLContext const&)+0x2aa) [0x152a86a2cd7a]
  [bt] (6) /home/user/anaconda3/envs/pgmg/lib/python3.8/site-packages/dgl/libdgl.so(dgl::HeteroGraph::CopyTo(std::shared_ptr<dgl::BaseHeteroGraph>, DLContext const&)+0xf5) [0x152a8692cfe5]
  [bt] (7) /home/user/anaconda3/envs/pgmg/lib/python3.8/site-packages/dgl/libdgl.so(+0x73a07b) [0x152a8693a07b]
  [bt] (8) /home/user/anaconda3/envs/pgmg/lib/python3.8/site-packages/dgl/libdgl.so(DGLFuncCall+0x48) [0x152a868bc378]`
and I managed to the epoch28. When I decided to generate the molecules I encountered another error 
`Traceback (most recent call last):
  File "generate.py", line 78, in <module>
    model, tokenizer = load_model(args.model_path, args.tokenizer_path)
  File "generate.py", line 34, in load_model
    print(model.load_state_dict(states['model'], strict=False))
  File "/home/user/anaconda3/envs/pgmg/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1667, in load_state_dict
    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for PGMG:
	size mismatch for pp_v_init.weight: copying a param with shape torch.Size([384, 9]) from checkpoint, the shape in current model is torch.Size([384, 8]).` which I think was related to the line23 in generate.py and I changed that to 8+1. However, when trying agian, another error appeared 
`<All keys matched successfully>
  0%|                                                                                  | 0/20 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "generate.py", line 94, in <module>
    res.extend(tokenizer.get_text(model.generate(g_batch)))
  File "/home/user/anaconda3/envs/pgmg/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/home/user/PGMG-main15/model/pgmg.py", line 279, in generate
    vv, vvs, pp_mask = self.process_p(pp_graphs)
  File "/home/user/PGMG-main15/model/pgmg.py", line 153, in process_p
    v = self.pp_v_init(pp_graphs.ndata['h'])
  File "/home/user/anaconda3/envs/pgmg/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/user/anaconda3/envs/pgmg/lib/python3.8/site-packages/torch/nn/modules/linear.py", line 114, in forward
    return F.linear(input, self.weight, self.bias)
RuntimeError: mat1 and mat2 shapes cannot be multiplied (6656x8 and 9x384)`
after checking scripts one by one, I realized that changing the maximum number of pharmacophores and the adding of anion is not an easy and simple thing to do by just modify a few parameters in certain scripts, it needs the overall changes to all related scripts, which is very hard for someone like me that not an expert in Python and not familiar with your code. Hope that PGMG can be updated to version2.0 with enhancements and more features in the future.

**Comment by e-yi** [comment]
@kolmorgan , I'm really sorry to hear that you've encountered frustration while tweaking the code. I appreciate your efforts in addressing some problems independently. If you're interested in proceeding, the current issue appears to be a mismatch in the node feature size of `pp_graph` during generation.

To address this, please follow these steps:

1. Add cation in `idx2phar` (L10 in `file_utils.py`)
2. Include cation size in `idx2size` (L19 in `file_utils.py`)  ( in fact, just add `7: ((1,), (1,)),` )
3. change `7` to  `len(idx2phar)` in L48(`tp = [0] * 7`) of `file_utils.py`

If you have any further questions or need assistance, feel free to reach out. 

P.S. Thanks for the suggestion! We're actively developing PGMG version 2.0 with the requested features.

**Comment by kolmorgan** [comment]
@e-yi Thank you for your kind help!  It worked out after I modified the code following your suggestion, although the structures of the generated molecules look similar to before (maybe related to the same training set). It's very kind of you and your team to help me with the modifications of the code. A little suggestion for better performance in another aspect of PGMG is the generation of multivalence ligands (some people call branched molecules) sharing the same pharmacophores for each branch. In fact, my target receptor is a multimeric protein with each branch of the branched ligand binding on the same monomers of the multimeric protein. For example, imagine that two or more aspirins are covalently linked via a linker on the phenyl ring. Therefore, for each branch of the ligand, the pharmacophore is the same. Actually, when I used 8 pharmacophores to generate the molecules, some of the generated molecules were branched molecules (symmetric). Such branched molecules were suitable for some multimeric proteins (or protein with repeated binding domain) in the body, and no one as far as I know was working on this aspect of molecular generation. I'm just sharing my own thoughts in the molecular generation. Hope that will help for the better performance of PGMG or other programs you are working on. Anyway, thanks for your help!
